DE everything but events working

rshorey opened the pull request:
Full rewrite, pretty much.

paultag commented on openstates/de/__init__.py:
```diff
@@ -46,10 +46,11 @@
     from billy.scrape.utils import url_xpath
-    sessions = url_xpath('http://legis.delaware.gov/',
+    url = "http://legis.delaware.gov/Legislature.nsf/\
+            7CD69CCAB66992B285256EE0005E0727/FC256764B3B3DCAE85257E0E005F9CD8"
```
```diff
+    url = "http://legis.delaware.gov/Legislature.nsf/\
+            7CD69CCAB66992B285256EE0005E0727/FC256764B3B3DCAE85257E0E005F9CD8"
```

Was this s'posed to be

```diff
+    url = "http://legis.delaware.gov/Legislature.nsf/"\
+          "7CD69CCAB66992B285256EE0005E0727/FC256764B3B3DCAE85257E0E005F9CD8"
```

Or:

```diff
+    url = ("http://legis.delaware.gov/Legislature.nsf/"
+           "7CD69CCAB66992B285256EE0005E0727/FC256764B3B3DCAE85257E0E005F9CD8")
```

paultag commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,329 @@
-            actor = actions_get_actor(action, bill['chamber'])
-            attrs = dict(actor=actor, action=action, date=date)
-            attrs.update(**self.categorizer.categorize(action))
-            bill.add_action(**attrs)
```
i```diff
+                bill.add_action(**attrs)
```

I'm seeing stuff like:

```
19         {
20             "absent": [
21                 "0"
22             ],
23             "action": "Passed by House of Representatives. Votes: Passed   37 YES         1 NO     3 NOT VOTING     0 ABSENT     0 VACANT",
24             "actor": "lower",
25             "date": 1427173200.0,
26             "no_votes": [
27                 "1"
28             ],
29             "not_voting": [
30                 "3"
31             ],
32             "related_entities": [],
33             "type": [
34                 "bill:passed"
35             ],
36             "vacant": [
37                 "0"
38             ],
39             "yes_votes": [
40                 "37"
41             ]
42         },
```

rshorey replied:
I pulled this code wholesale from the old bathawk scraper in the hopes of keeping things consistent. Do you want me to take the votes out of actions (they should be in votes, too). The vote is listed as an action separately from votes on the state's site.

mileswwatkins replied:
I agree that this looks a bit funky, although I can see why it wouldn't necessarily live in `vote`s, especially if it's a vote _on_ an amendment, as opposed to on the bill itself. Personally, I'd just scrape the action text (with the vote totals) and leave it at that, not interpret it further.

rshorey replied:
Removed additional information from vote action (the counts are still in the text, since that's how we get it from the state, and the votes are still available as votes)

mileswwatkins commented on openstates/de/legislators.py:
```diff
@@ -47,19 +70,20 @@ def scrape(self, chamber, term, text=methodcaller('text_content'),
             party = 'Democratic'
         elif '(R)' in party:
             party = 'Republican'
+        print party
```
Debug code?

mileswwatkins commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,329 @@
-        for format_ in formats:
-
-            el =_doc.xpath('//font[contains(., "%s%s")]' % (filename, format_))
+            bill = Bill(session,chamber,bill_id,bill_title)
```
At this point for some `bill`s there's still some extra cruft (eg, newline and two `nbsp`s, or a leading space) on their `bill_id`. I can't quickly tell you where this isn't getting stripped, but here are some examples of `bills` with this problem: HB 18, HB 26, HB 37, SB 13, SB 16.

rshorey replied:
fixed spacing issue, found issue with duplicate versions from substitute bill, fixed that, too.

mileswwatkins commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,329 @@
-            totals = re_totals.findall(doc.text_content())
-            if len(totals) == 4:
-                self.warning('...was able to parse vote tallies from "%s"' % url)
+            self.logger.warning("Can't find bill number, skipping")
```
Seems like we get a slew of these errors when the scraper hits nominations. Could we `self.info()` skip instead of `self.warning()` if the scraper sees a nomination? Maybe if it sees "Nomination Information" as the page header, [like here](http://legis.delaware.gov/LIS/lis148.nsf/ac4d906954d0945685257e0b00557ac0/1c8ede0fde8f351485257dda005ba189?OpenDocument)?

rshorey replied:
added nominee-specific check. also I think I've finally learned to spell Delaware correctly consistently.

mileswwatkins commented on openstates/de/committees.py:
```diff
@@ -22,29 +22,41 @@ def scrape(self, chamber, term):
+
+
+
+
```
Space: The Final Fronteir

rshorey replied:
blank space removed.

mileswwatkins commented on openstates/de/committees.py:
```diff
@@ -22,29 +22,41 @@ def scrape(self, chamber, term):
 
         url = urls[chamber] % (session,)
         self.log(url)
```
Remove old debug code.

mileswwatkins commented on openstates/de/committees.py:
```diff
@@ -22,29 +22,41 @@ def scrape(self, chamber, term):
+        page = lxml.html.fromstring(self.get(url).text)
         page.make_links_absolute(url)
 
         committees = {}
```
Drop variable; no longer used, and might be confusing for later upkeepers.

rshorey replied:
variable dropped.

mileswwatkins commented on openstates/de/committees.py:
```diff
@@ -22,29 +22,41 @@ def scrape(self, chamber, term):
+            if len(row.xpath('./td')) > 0:
+                #if statement removes header tr
+                comm = row.xpath('.//a')[1]
+                comm_name = comm.text_content().strip()
```
All of the committee names lack "Committee" in this form, with the exception of [House Sunset Committee](http://legis.delaware.gov/LIS/LIS148.NSF/*/B1635714074F9E3085257DC400560ECB/?opendocument&nav=House). And I think that's just because there's a weird parenthetical in the name, immediately after "Committee." How would you feel about standardizing that name by removing "Committee?"

rshorey replied:
This is how they show up on the source (nothing has "committee" except that one) so I'm leaving it.

mileswwatkins commented on openstates/de/committees.py:
```diff
@@ -22,29 +22,41 @@ def scrape(self, chamber, term):
 
         url = urls[chamber] % (session,)
```
What about the [Joint Finance Committee](http://www.legis.delaware.gov/legislature.nsf/Lookup/CGO_JFCDescription), which I think is the only joint committee in the state legislature? It's partially scraped as a [House "committee,"](http://legis.delaware.gov/LIS/LIS148.NSF/*/D4ABF3766FDC0A8485257DC400542E52/?opendocument&nav=House) but it'd be nice to make this a full scrape as `joint` and throw out the partial House scrape.

mileswwatkins replied:
Actually, there is [at least one more `joint` committee](http://legis.delaware.gov/legislature.nsf/lookup/Bond_home?open). And [according to Ballotpedia](http://ballotpedia.org/Delaware_General_Assembly#Joint_committees) the Sunset Committee is actually `joint`, although we currently successfully scrape it separately for `upper` and `lower`.

rshorey replied:
Joint committees in DE are really confusing. I have a call in to their leg to figure this out, and am going to put this on hold till they get back to me.

mileswwatkins commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,329 @@
-                _kwargs['filename'] = _kwargs['filename'].lower()
 
-            url = tmp.format(**_kwargs).replace(' ', '+')
 
```
Lop off some of these empty lines at the EOF.

rshorey replied:
empty space gone.

mileswwatkins commented on openstates/de/legislators.py:
```diff
@@ -16,27 +16,50 @@ def scrape(self, chamber, term, text=methodcaller('text_content'),
+            name_and_url = tr.xpath('.//a')[0]
+            bio_url = name_and_url.attrib["href"]
+            bio_url = bio_url.replace("JavaScript:window.location.href=","")
+            bio_url = base_url + bio_url.replace('"','')
```
There's double-`/`s in this, since both `base_url[-1]` and `bio_url[0]` are `/`s.

rshorey replied:
double slashes fixed

mileswwatkins commented on openstates/de/legislators.py:
```diff
@@ -68,64 +92,64 @@ def scrape_bio(self, term, chamber, district, name, url):
-                info['email'] = email.strip()
-                break
+        email = doc.xpath(".//a[contains(@href,'mailto')]")
+        email = email[0].text_content().strip()
```
Sometimes this `text_content()` will capture two email addresses, not just one. For example, [Sokola](http://legis.delaware.gov//legislature.nsf/vwMain/E5D26F90AC51457285257E0E005A56C6?Opendocument). This seems to be remedied if you take the `@href` instead.

rshorey replied:
If there are two e-mail addresses, any reason not to add the second to the leg's "outside office"?

rshorey replied:
OK, I've done something reasonable with e-mail addresses.

mileswwatkins commented on openstates/de/legislators.py:
```diff
@@ -16,27 +16,50 @@ def scrape(self, chamber, term, text=methodcaller('text_content'),
                re_spaces=re.compile(r'\s{,5}')):
```
Remove the non-standard parameters to this function (`text` and `re_spaces`). This should also allow you to remove some unused dependencies. @paultag suggested that these parameters are probably from someone trying to run the scraper outside of `billy`, which is a really odd use-case.

rshorey replied:
right, weird, parameters and dependencies fixed.

mileswwatkins commented on openstates/de/legislators.py:
```diff
@@ -16,27 +16,50 @@ def scrape(self, chamber, term, text=methodcaller('text_content'),
 
-        doc = lxml.html.fromstring(self.urlopen(url))
+        doc = lxml.html.fromstring(self.get(url).text)
         doc.make_links_absolute(url)
```
Might as well `lxmlize` as long as you're changing this.

rshorey replied:
legislators lxmlized everywhere.

mileswwatkins commented on openstates/de/legislators.py:
```diff
@@ -47,19 +70,20 @@ def scrape(self, chamber, term, text=methodcaller('text_content'),
         if '(D)' in party:
             party = 'Democratic'
         elif '(R)' in party:
             party = 'Republican'
```
Error if neither found?

rshorey replied:
Error now thrown if no party found.

mileswwatkins commented on openstates/de/legislators.py:
```diff
@@ -68,64 +92,64 @@ def scrape_bio(self, term, chamber, district, name, url):
+        leg_office = dict(name="Capitol Office", type="capitol",
+                        phone=None, fax=None, email=email, address=None)
+        dist_office = dict(name="District Office", type="district",
+                        phone=None,fax=None, email=email, address=None) 
```
Do we duplicate the assignment of `phone`s or `email`s to multiple offices? I've never really considered this before.

rshorey replied:
I've thought this through for the guy with two e-mail addresses and am now doing something reasonable.

mileswwatkins commented on openstates/de/legislators.py:
```diff
@@ -68,64 +92,64 @@ def scrape_bio(self, term, chamber, district, name, url):
+
+            if "outside office" in title_text:
+                dist_office["address"] = content.strip()
+            if "outside phone" in title_text:
```
AFAICT, this is identical to the `"legislative phone"` code; could it be put into a dedicated function?

rshorey replied:
Legislative offices DRY'd

mileswwatkins commented on openstates/de/legislators.py:
```diff
@@ -68,64 +92,64 @@ def scrape_bio(self, term, chamber, district, name, url):
+                        elif "fax" in line:
+                            dist_office["fax"] = line.replace("fax:","").strip()
+
+        offices = [o for o in [leg_office,dist_office] if o["address"]]
```
Maybe `assert` that at least one office has an `address`? Otherwise, we may be throwing out useful email and phone data.

rshorey replied:
at least one address asserted.

mileswwatkins commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,339 @@
+                                    person_seen = True
+
+            if voice_vote:
+                vote = Vote(vote_chamber,date,"passage",passed,0,0,0)
```
@paultag, do we treat voice votes as `vote`s? Eg,
http://legis.delaware.gov/LIS/lis148.nsf/7712cf7cc0e9227a852568470077336f/8cedd7e55eba128485257ddc007b3fc6?OpenDocument

rshorey replied:
voice votes are being counted as votes, confirmed as OK.

mileswwatkins commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,339 @@
+            if "long title" in title_text:
+                bill_title = tds[1].text_content().strip()
+
+            if "amendment" in title_text:
```
Sometimes amendment webpages have their own `vote`s on them, that should be scraped accordingly. For example, [SB 10](http://legis.delaware.gov/LIS/lis148.nsf/ac4d906954d0945685257e0b00557ac0/94b6981d895a153785257dcb005ca435?OpenDocument) has [an amendment](http://www.legis.delaware.gov/LIS/lis148.nsf/vwLegislation/SA+1+to+SB+10?opendocument) with [a `vote`](http://www.legis.delaware.gov/LIS/lis148.nsf/7712cf7cc0e9227a852568470077336f/73f0925a5c34894685257dd5007173cc?OpenDocument) that is not currently scraped.

rshorey replied:
messy. but amendment votes being added now.

mileswwatkins commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,339 @@
+        for bill_code in bill_codes[chamber]:
+            page = self.lxmlize(base_url.format(session=session,
+                                            bill_type=bill_code,
+                                            count=1000)) #1000 shld be enough
```
Maybe note that previous sessions in the Open States database haven't had more than 1,000 bills _total_.

rshorey replied:
comment updated

mileswwatkins commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,339 @@
+                    add_spons_text = add_spons_text.replace("Additional Sponsor(s):","")
+                    addl_sponsors = self.separate_names(add_spons_text)
+
+            if "co-sponsor" in title_text:
```
Make these `elif`s, since they're mutually exclusive?

rshorey replied:
ifs elif'd

mileswwatkins commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,339 @@
+
+            if "co-sponsor" in title_text:
+                cosponsor_text = tds[1].text_content()
+                if "none" in cosponsor_text.lower():
```
Can this conditional be made more specific? Would want to guard against odd names. Maybe `r'(?i)(?<![A-z])none(?!.*[A-z])'`?

rshorey replied:
Conditional more specific to avoid weird names. But we're going to be in trouble when Joe Cosponsor gets elected somewhere.

mileswwatkins commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,339 @@
+                if line.strip().startswith("Date"):
+                    date_str = " ".join(line.split()[1:4])
+                    date = datetime.strptime(date_str,"%m/%d/%Y %I:%M %p")
+                    passed = "Passed" in line
```
Is "Passed" the only keyword for passage? From the pages I've looked at it is, but maybe we could check the last bold word in `line`, and `assert` that it's in a list of accepted statuses?

rshorey replied:
We've never seen a status besides "passed". I added an error if we ever do so we can figure out what to do with it in that case.

mileswwatkins commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,339 @@
+                        voice_vote = True
+                    else:
+                        voice_vote = False
+                        yes_count = int(re.findall("Yes: (\d*)",line)[0])
```
`{1,2}`, or at least `+`, instead of `*`?

rshorey replied:
Changing to + because of a strong preference for minimal regular expressions and the knowledge that if we actually catch anything incorrect it'll get found in later validations :)

mileswwatkins commented on openstates/de/bills.py:
```diff
@@ -1,693 +1,339 @@
+                vote["no_votes"] = no_votes
+                vote["other_votes"] = other_votes
+
+            assert len(vote["yes_votes"]) == vote["yes_count"], \
```
A call to `vote.validate()` can replace these three checks.

rshorey replied:
didn't we decide these get called when the vote is saved? I'll just delete the asserts.

mileswwatkins replied:
I think you're right? But I do like having `vote` validation in at `scrape`-time. Either way should be totally fine.

rshorey commented:
I've addressed all of the comments on DE. The diff is now a mess, I'm sorry, this was a bear of a site to wrangle. Thanks for the review, and let me know what else you see.

mileswwatkins commented:
Looks good to go, otherwise!

rshorey commented:
@mileswwatkins, all addressed. ran locally without cache and it didn't crash.

mileswwatkins commented:
Ditto.
